service/history/api/multioperation/api.go

```diff
@@ -128,40 +128,47 @@ func Invoke(
 }
 
 func (mo *multiOp) Invoke(ctx context.Context) (*historyservice.ExecuteMultiOperationResponse, error) {
-	// For workflow id conflict policy terminate-existing, always attempt a start
-	// since that works when the workflow is already running *and* when it's not running.
 	conflictPolicy := mo.startReq.StartRequest.WorkflowIdConflictPolicy
-	if conflictPolicy == enumspb.WORKFLOW_ID_CONFLICT_POLICY_TERMINATE_EXISTING {
-		resp, err := mo.startAndUpdateWorkflow(ctx)
-		var noStartErr *noStartError
-		switch {
-		case errors.As(err, &noStartErr):
-			// The start request was deduped, no termination is needed.
-			// Continue below by only sending the update.
-		case err != nil:
-			return nil, err
-		default:
-			return resp, nil
-		}
-	}
 
-	runningWorkflowLease, err := mo.getRunningWorkflowLease(ctx)
+	workflowLease, err := mo.getWorkflowLease(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	// Workflow was already started ...
-	if runningWorkflowLease != nil {
-		if err = mo.allowUpdateWorkflow(ctx, runningWorkflowLease, conflictPolicy); err != nil {
-			runningWorkflowLease.GetReleaseFn()(nil) // nil since nothing was modified
-			return nil, err
+	// Workflow was started already.
+	if workflowLease != nil {
+		// Workflow is still running and not requested to be terminated - send the update.
+		if workflowLease.GetMutableState().IsWorkflowExecutionRunning() && conflictPolicy != enumspb.WORKFLOW_ID_CONFLICT_POLICY_TERMINATE_EXISTING {
+			if err = mo.allowUpdateRunningWorkflow(workflowLease, conflictPolicy); err != nil {
+				workflowLease.GetReleaseFn()(nil) // nil since nothing was modified
+				return nil, err
+			}
+			return mo.updateWorkflow(ctx, workflowLease) // lease released inside
+		}
+
+		// Workflow is not running anymore and the update already completed - return the outcome.
+		updateId := mo.updateReq.Request.Request.Meta.GetUpdateId()
+		if outcome, err := workflowLease.GetMutableState().GetUpdateOutcome(ctx, updateId); err == nil {
+			workflowKey := workflowLease.GetContext().GetWorkflowKey()
+			workflowLease.GetReleaseFn()(nil)
+
+			return makeResponse(
+				&historyservice.StartWorkflowExecutionResponse{
+					RunId:   workflowKey.RunID,
+					Started: false, // set explicitly for emphasis
+					// TODO: Running: false,
```

Comment:
This requires an API change.

Reply:
I guess you need to rename it to `status` here too but as soon as it is just a comment it doesn't really matter.

Reply:
👍 I'll save the planet by not pushing the comment update now; and will fix it in the next PR.